### DISCUSSION TITLE
Loading behavior of `app/overrides` path in `application.rb`

### DISCUSSION THREAD

**Discussion by rabbitbike** [solidusio/solidus#5349] [Troubleshooting] [answered]
Hi,

I and my co-developer are currently trying to upgrade our store from 3.1 to 3.4, hoping to go all the way up to 4.1 soon.
We are excited about the update, especially the new Admin UI. We want to try it as soon as we can!

According to the [v3.2 documentation](https://solidus.io/blog/solidus-v3-2), we should stop using `app/**/*_decorator*.rb` files and place renamed files in `app/overrides` directory. Once the files are there, we should put the following in the `application.rb`.

```
overrides = "#{Rails.root}/app/overrides" # use your actual directory here
Rails.autoloaders.main.ignore(overrides)
config.to_prepare do
  Dir.glob("#{overrides}/**/*_decorator*.rb").each do |override|
    load override
  end
end
```

I notice that when I do this, the files in the directory are loaded twice and result in warnings and errors.
For example, a constant gets a warning stating that it has been defined already.
Also, some classes from extensions don't seem to be loaded yet; for example, I get an error for not finding `Spree::Taxon::Translation` which is from `solidus_globalize` gem.

Once I remove the above code, rails server starts without an issue and all files in the `app/overrides` are correctly loaded. So it seems like the files in `app/overrides` path are actually loaded without our having to load them specifically.

I looked around for the reason for this behavior, and it seems like the deface gem already puts this path in autoloading. ([The code maybe here?](https://github.com/spree/deface/blob/master/lib/deface/environment.rb))

I found a workaround for this, and it is to use a name other than `overrides.` Then things work as expected with the above code.
I understand that the name of the directory can be something other than 'overrides,' but I would like to keep the recommended naming. Since it's my understanding that deface is deeply integrated with Solidus, I am confused about this behavior and I would appreciate it if documentation gives us some clue.

~Should I just comment out this part of the code in `application.rb`?~
~Or is this something we should change after we migrate to Rails 7?~
(Update: See answer below)

I am working on a branch that uses solidus 3.4 gems.
`Spree.load_defaults '3.4'` is turned on.
Ruby is 3.2.2 and Rails is 6.1.7. The autoloader is Zeitwerk.

Thank you for your help🙏

**Comment by waiting-for-dev** [comment, marked as answer]
Hi @rabbitbike, all your assumptions are correct. You were facing the issue because of Deface, and using another directory will fix your warnings. Adding a warning in that post would be good, as I'm sure others will find the same issue. We're building the new admin with customization at its base, so we expect Deface shouldn't be needed anymore.

**Reply by rabbitbike** [reply]
Ok, we will use another directory.
I'm excited to learn that the planned new admin customization will do away with Deface.
Yes, I think it would be helpful for those making the changes if the documentation mentions this somewhere.

Thank you for clarifying the issue and we really appreciate the Solidus team's work🙏

**Reply by rabbitbike** [reply]
Also, not loading the files in the `application.rb` and relying on Deface's autoloading of the `app/overrides` will cause loading order issues for some engines such as solidus_globalize with added classes not being loaded in time.
So the correct way is to use a directory with another name as answered and include the loading code in `application.rb`.

**Reply by waiting-for-dev** [reply]
FYI, I added a notice to the blog post, and I also updated the Changelog in this repo (see #5352). Thanks a lot for reporting the issue.
